src/pools/D3MCompoundPool.sol

```diff
@@ -0,0 +1,200 @@
+    function _wmul(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        z = (x * y) / WAD;
+    }
+    function _wdiv(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        z = (x * WAD) / y;
+    }
+    function _min(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        z = x <= y ? x : y;
+    }
+
+    // --- Admin ---
+    function rely(address usr) external auth {
+        wards[usr] = 1;
+        emit Rely(usr);
+    }
+    function deny(address usr) external auth {
+        wards[usr] = 0;
+        emit Deny(usr);
+    }
+
+    function file(bytes32 what, address data) external auth {
+        require(vat.live() == 1, "D3MCompoundPool/no-file-during-shutdown");
+        if (what == "hub") {
+            vat.nope(hub);
+            hub = data;
+            vat.hope(data);
+        } else if (what == "king") king = data;
+        else revert("D3MCompoundPool/file-unrecognized-param");
+        emit File(what, data);
+    }
+
+    function deposit(uint256 wad) external override onlyHub {
+        uint256 prev = cDai.balanceOf(address(this));
+        require(cDai.mint(wad) == 0, "D3MCompoundPool/mint-failure");
+        // As interest was accrued on `mint` we can use the non accruing `exchangeRateStored`
+        require(
+            cDai.balanceOf(address(this)) ==
+            prev + _wdiv(wad, cDai.exchangeRateStored()), "D3MCompoundPool/incorrect-cdai-credit"
+        );
+    }
+
+    function withdraw(uint256 wad) external override onlyHub {
+        require(cDai.redeemUnderlying(wad) == 0, "D3MCompoundPool/redeemUnderlying-failure");
+        dai.transfer(msg.sender, wad);
+    }
+
+    function transfer(address dst, uint256 wad) external override onlyHub {
+        require(cDai.transfer(dst, _wdiv(wad, cDai.exchangeRateCurrent())), "D3MCompoundPool/transfer-failed");
+    }
+
+    function quit(address dst) external override auth {
+        require(vat.live() == 1, "D3MCompoundPool/no-quit-during-shutdown");
+        require(cDai.transfer(dst, cDai.balanceOf(address(this))), "D3MCompoundPool/transfer-failed");
+    }
+
+    function preDebtChange(bytes32) external override {
+        require(cDai.accrueInterest() == 0, "D3MCompoundPool/accrueInterest-failure");
+    }
+
+    function postDebtChange(bytes32) external override {}
+
+    // Does not accrue interest (as opposed to cToken's balanceOfUnderlying() which is not a view function).
+    function assetBalance() public view override returns (uint256) {
+        (uint256 error, uint256 cTokenBalance,, uint256 exchangeRate) = cDai.getAccountSnapshot(address(this));
+        require(error == 0, "D3MCompoundPool/getAccountSnapshot-failure");
+        return _wmul(cTokenBalance, exchangeRate);
+    }
+
+    function maxDeposit() external pure override returns (uint256) {
+        return type(uint256).max;
+    }
+
+    function maxWithdraw() external view override returns (uint256) {
+        return _min(cDai.getCash(), assetBalance());
```

Comment:
I was wondering if the reserves amount could influence the max amount to withdraw but from what I followed in the code, whatever DAI is available as cDai balance can be withdrawn. So all should be fine.
Just wanted to make sure this was also checked from your side and we can close this comment.

Reply:
Yeah that's right.
In any given time:
total amount of withdrawal rights = cash + borrows - reserves.

So if you withdraw the entire balance (`cash`) it just means that the reserves are borrowed out.
But still after you do that the amount of borrows will be the total amount of withdrawal rights + reserves.